### PR TITLE
feat: adding action yaml and readme

### PR DESCRIPTION
**Pull request by johnson2427**

### CODE REVIEW

**Comment by johnson2427** [comment]
@fubuloubu I made this very strict. Essentially, don't deviate from silverback's suggestions. Let me know if you want to free this up a bit. It'll require quite a bit of rework to allow different directories and file names. We can definitely do it, just won't be able to do so with the current `silverback` version. We can either make `silverback build` really smart, or we can just deal with inputs here. Up to you.

File: .github/workflows/.gitkeep
Comment (fubuloubu) [inline .github/workflows/.gitkeep:1]: We can probably test by packaging up the example app. I wanted that anyways, based on docs how to run another published silverback container image

File: README.md [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
+
+## Outputs
+
+### `build_images`
+The images that are built and pushed by the workflow
+
+## Example Usage
+```
+steps:
+  - name: silverback-build-push
```
Comment (fubuloubu) [inline README.md]:
```suggestion
  - name: Build image
```

File: action.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+          exit 1
+        fi
+
+    - name: Log into GitHub Container Registry
+      if: github.event_name != 'pull_request'
+      uses: docker/login-action@v2
+      with:
+        registry: ${{ env.REGISTRY }}
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
```
Comment (fubuloubu) [inline action.yml]: I was thinking that the login and publish part could be done separately. It should at the very least be user configurable when to publish.

File: .github/workflows/.gitkeep
Comment (fubuloubu) [reply .github/workflows/.gitkeep:1]: I suppose once this is out there, we can just test it against the live silverback repo.

Comment (johnson2427) [reply .github/workflows/.gitkeep:1]: I will add an example app for this to test

File: action.yml (hunk 2 above)
Comment (johnson2427) [reply action.yml]: Agreed, and then I can allow for configure on push. I've been working through as many ideas I can think of. i think I have a good option here. Should be out by eod

File: .github/workflows/.gitkeep
Comment (fubuloubu) [reply .github/workflows/.gitkeep:1]: no I mean specifically this one:
https://github.com/ApeWorX/silverback/blob/main/example.py

e.g. `ghcr.io/apeworx/silverback-example`

File: action.yml (hunk 2 above)
Comment (fubuloubu) [reply action.yml]: Yeah, I think if it's an arg to this function (defaults false), you can just use like `if: ${{ arg.PUSH == true }}` or something

File: README.md [hunk 3]
```diff
@@ -1,2 +1,15 @@
...
+## Outputs
+
+### `build_images`
+The images that are built and pushed by the workflow
+
+## Example Usage
+```
+steps:
+  - name: Build Image
+    uses: SilverbackLtd/silverbackltd-build@v3
```
Comment (fubuloubu) [inline README.md]: Use the current name, and we should release a v1 tag first
```suggestion
    uses: SilverbackLtd/build-action@v1
```

File: action.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
+name: 'silverbackltd-build-action'
```
Comment (fubuloubu) [inline action.yml]: Can this have spaces in it?
```suggestion
name: 'Silverback Build and Push Images'
```

File: action.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  icon: 'check-circle'
+  color: 'green'
+
+inputs:
+  push:
+    description: 'Push the image to ghcr'
+    required: false
+    default: 'false'
+
+outpus:
```
Comment (fubuloubu) [inline action.yml]:
```suggestion
outputs:
```

File: action.yml [hunk 6]
```diff
@@ -0,0 +1,91 @@
+name: 'silverbackltd-build-action'
+author: 'Silverback Ltd.'
```
Comment (fubuloubu) [inline action.yml]:
```suggestion
author: 'ApeWorX Ltd.'
```

File: action.yml [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+description: 'A GitHub Action to build docker images for a silverback repository and push to a container registry'
```
Comment (fubuloubu) [inline action.yml:3]:
```suggestion
description: 'A GitHub Action to build docker images for all Silverback bots in a repository and push to a container registry'
```

File: action.yml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    description: 'List of Docker images that were built and pushed.'
+
+runs:
+  using: 'composite'
+  steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+
```
Comment (fubuloubu) [inline action.yml]: Don't think you need this, suggest to put into docs as something the user will do anyways

Comment (fubuloubu) [reply action.yml]: Also is fetch depth 0 required?

File: action.yml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        registry: ${{ env.REGISTRY }}
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Build and Push
+      run: |
+        # Find all Dockerfiles matching the pattern
+        dockerfiles=$(find "${{ env.TARGET_DIR }}" -type f -name "${{ env.DOCKERFILE_PATTERN }}" | sort)
+        echo "Found Dockerfiles:"
+        echo "${dockerfiles}"
```
Comment (fubuloubu) [inline action.yml:102]: I feel like the repo tag should be an input defaulted to `latest`

File: README.md [hunk 10]
```diff
@@ -1,2 +1,15 @@
 # silverbackltd-build
```
Comment (fubuloubu) [inline README.md]: I believe this gets published to the Actions store page
```suggestion
 # Silverback Build Workflow
```

File: README.md [hunk 11]
```diff
@@ -1,2 +1,15 @@
...
-SilverbackLTD Build workflow
+SilverbackLTD Build workflow will automatically generate dockerfiles, build docker images and push those images to ghcr.
```
Comment (fubuloubu) [inline README.md]:
```suggestion
Automatically generate dockerfiles (unless already commited), build docker images and push those images to a configured container registry like GHCR
```

File: action.yml (hunk 8 above)
Comment (johnson2427) [reply action.yml]: this is not needed, hold over from an old repository

File: action.yml (hunk 9 above)
Comment (johnson2427) [reply action.yml:102]: agreed

File: .github/workflows/.gitkeep
Comment (johnson2427) [reply .github/workflows/.gitkeep:1]: oh, I'm not sure how to do that at the moment. searching

File: README.md [hunk 12]
```diff
@@ -1,2 +1,42 @@
...
+# Silverbackltd Build Workflow
+
+Automatically generate dockerfiles (unless already committed), build docker images and push those images to a configured container registry like GHCR
+
+## Outputs
+
+### `build_images`
+The images that are built and pushed by the workflow
+
+## Example Usage
```
Comment (fubuloubu) [inline README.md:10]: Nice, this looks perfect UX

File: README.md [hunk 13]
```diff
@@ -1,2 +1,42 @@
...
+
+To push to a ghcr registry:
```
Comment (fubuloubu) [inline README.md:12]: make sure to include permissions required

File: action.yml [hunk 14]
```diff
@@ -0,0 +1,106 @@
+name: Silverbackltd Build Action
+author: 'ApeWorX Ltd.'
+description: 'A GitHub Action to build docker images for a silverback repository and push to a container registry'
+branding:
+  icon: 'check-circle'
+  color: 'green'
+
+inputs:
+  push:
+    description: 'Push the image to ghcr'
```
Comment (fubuloubu) [inline action.yml]:
```suggestion
    description: 'Push the image to the configured registry'
```

File: action.yml [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    required: false
+    default: 'latest'
+  registry:
+    description: 'The registry location you want to push to.'
+    required: false
+    default: 'ghcr.io'
+  username:
+    description: 'Add your username for your registry'
+    required: false
+    default: ${{ github.actor }}
```
Comment (fubuloubu) [inline action.yml]: Does this work?

File: action.yml [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  password:
+    description: 'Add your password for your registry'
+    required: false
+    default: ''
```
Comment (fubuloubu) [inline action.yml]: This shouldn't work with default, but then again it's probably not required unless you push right?

File: ape-config.yaml
Comment (fubuloubu) [inline ape-config.yaml:1]: The silverback build command should generate a dockerfie w/ `COPY ape-config.yaml .` if and only if this file exists

Can you update SDK?

File: requirements.txt
Comment (fubuloubu) [inline requirements.txt:1]: The silverback build command should generate a dockerfie w/ `COPY ape-config.yaml .` if and only if this file exists

Can you update SDK?

**Review by fubuloubu — COMMENTED** [review]
A few things, but getting close to ready!

File: ape-config.yaml
Comment (johnson2427) [reply ape-config.yaml:1]: yes, now that this works, with the gross workarounds, I'll make those changes

File: action.yml (hunk 16 above)
Comment (johnson2427) [reply action.yml]: correct

File: action.yml (hunk 15 above)
Comment (johnson2427) [reply action.yml]: need to test an actual push, I doubt it. I should probably `default: ''` like I did with password, honestly. We don't need it unless we push

File: README.md (hunk 13 above)
Comment (johnson2427) [reply README.md:12]: ah, good call

File: action.yml (hunk 16 above)
Comment (johnson2427) [reply action.yml]: I should add a check at the beginning of the job to check if push is true, then confirm there are values there. If there aren't give a descriptive error

**Review by fubuloubu — APPROVED** [review]
Good enough to try some testing on repos. I think once the fix is available from silverback for ape-config.yaml/requirements.txt not being there, and we have done a little testing, this is good to go

**Pull request merged**